### PR TITLE
GeoJSONLayer Enhancements

### PR DESCRIPTION
**Pull request by daniel5151**
I found the GeoJSONLayer component to be a bit under-featured, so I enhanced it a bit:

- Added support for mapbox's fill-extrusion layer (for displaying 3D buildings and such)
- Updated `componentWillReceiveProps` to update the map on any changes to the `[layer]Paint` and `[layer]Layout` props
- Changed it so that by default, GeoJSONLayer will not render any layers that do not have explicitly defined Paint / Layout props

This resolves issue #144 

(Please ignore that last PR, I committed with the wrong account 😅. This is my first PR, cut me some slack 😄)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -62,6 +62,7 @@
     "react-dom": "^15.0.1"
   },
   "devDependencies": {
+    "@types/core-js": "^0.9.36",
```
Comment (alex3165) [inline package.json]: I don't think this is needed

File: src/geojson-layer.ts [hunk 2]
```diff
@@ -1,25 +1,39 @@
 import * as React from 'react';
 import * as MapboxGL from 'mapbox-gl/dist/mapbox-gl';
+const isEqual = require('deep-equal');
+import diff from './util/diff';
 import { generateID } from './util/uid';
 import { Sources, SourceOptionData } from './util/types';
 
+const type_to_layer_LUT = {
```
Comment (alex3165) [inline src/geojson-layer.ts]: Can you use camel case ?

File: package.json (hunk 1 above)
Comment (daniel5151) [inline package.json]: `npm run build` would fail for me until I added this dep.

**Comment by alex3165** [comment]
Awesome 👍  do you think you could write some test for geojson-layer as well ? if not it is fine we can first merge this and add some tests later

**Comment by daniel5151** [comment]
Tests would be good, but I don't really have the time to do them now. Sorry 😞

**Comment by alex3165** [comment]
Also can you remove the yarn.lock ? don't think this should go out on this PR, it can be added separately later

File: package.json (hunk 1 above)
Comment (alex3165) [inline package.json]: Ok fair enough 👍

**Comment by alex3165** [comment]
I have checked the GeoJson example on your branch and there is an issue the points doesn't appear on the map anymore.

**Comment by daniel5151** [comment]
Ah, yes, the example... I forgot about that.
Off the top of my head, the reason they are not showing up is because I changed the default behaviour of the component to not render any layers that have not been given Paint/Layout props.

I'll update the example to show off those hidden layers too.

File: docs/API.md [hunk 3]
```diff
@@ -140,8 +140,8 @@ import { GeoJSONLayer } from "react-mapbox-gl";
 
 ### Properties
 - **data** *(required)* : `String | Object` The url to the geojson file or the geojson file itself.
-- **lineLayout** | **symbolLayout** | **circleLayout** | **fillLayout** : `Object` Layer layout informations. [mapbox layout api](https://www.mapbox.com/mapbox-gl-style-spec/#layer-layout)
-- **linePaint** | **symbolPaint** | **circlePaint** | **fillPaint** : `Object` Paint informations. [mapbox paint api](https://www.mapbox.com/mapbox-gl-style-spec/#layer-paint)
+- **lineLayout** | **symbolLayout** | **circleLayout** | **fillLayout** | **fillExtrusionLayout** : `Object` Layer layout informations. [mapbox layout api](https://www.mapbox.com/mapbox-gl-style-spec/#layer-layout)
```
Comment (philpl) [inline docs/API.md]: (Unrelated to this line)

can you remove the example `yarn.lock` file? I know it's unnecessary and we all love yarn, but unless the entire project supports it, we might not want to create any confusion ;)

File: src/geojson-layer.ts [hunk 4]
```diff
@@ -1,25 +1,39 @@
 import * as React from 'react';
 import * as MapboxGL from 'mapbox-gl/dist/mapbox-gl';
+const isEqual = require('deep-equal'); //tslint:disable-line
```
Comment (philpl) [inline src/geojson-layer.ts]: @alex3165 Are there typings for deep equal? if not, can we add some?

File: src/geojson-layer.ts [hunk 5]
```diff
@@ -50,8 +64,11 @@ export default class GeoJSONLayer extends React.Component<Props, void> {
     const layerId = `${id}-${type}`;
     layerIds.push(layerId);
 
-    const paint: Paints = this.props[`${type}Paint`] || {};
-    const layout: Layouts = this.props[`${type}Layout`] || {};
+    const paint: Paints = this.props[`${typeToLayerLUT[type]}Paint` ] || {};
+
+    // deafult undefined layers to invisible
```
Comment (philpl) [inline src/geojson-layer.ts]: small typo

File: docs/API.md (hunk 3 above)
Comment (alex3165) [inline docs/API.md]: He already removed the yarn.lock file

File: src/geojson-layer.ts (hunk 4 above)
Comment (alex3165) [inline src/geojson-layer.ts]: no typings for deep-equal indeed, we could write some or use `equal.lodash` instead maybe

File: docs/API.md (hunk 3 above)
Comment (alex3165) [inline docs/API.md]: Actually it is here again, @daniel5151 please add it to the `.gitignore`

**Comment by daniel5151** [comment]
That's weird... why is Travis failing? I don't think this latest commit should have affected anything.

**Comment by alex3165** [comment]
@daniel5151 Something is wrong in the latest node typings file:
```
node_modules/@types/node/index.d.ts(50,14): error TS2687: All declarations of 'iterator' must have identical modifiers.
```

Can you try an older version of typings for node ?

**Comment by daniel5151** [comment]
That seems to have done the trick

File: src/geojson-layer.ts [hunk 6]
```diff
@@ -50,8 +64,11 @@ export default class GeoJSONLayer extends React.Component<Props, void> {
     const layerId = `${id}-${type}`;
     layerIds.push(layerId);
 
-    const paint: Paints = this.props[`${type}Paint`] || {};
-    const layout: Layouts = this.props[`${type}Layout`] || {};
+    const paint: Paints = this.props[`${typeToLayerLUT[type]}Paint` ] || {};
+
+    // default undefined layers to invisible
+    const visibility = Object.keys(paint).length ? 'visible' : 'none';
```
Comment (alex3165) [inline src/geojson-layer.ts]: Hey, having a second look at the PR and I am wondering, why do you default to invisible ? Can you convince me it is an expected behaviour by default ?

Comment (alex3165) [inline src/geojson-layer.ts]: Ok I get it, can you please explain the new behaviour in the description of the `GEOJson` component on the documentation ?

**Comment by philpl** [comment]
@alex3165 I'd say this is ready to go for now?

**Comment by alex3165** [comment]
Thanks @daniel5151 nice job and thanks @philpl for the review 👍

**Pull request merged**